boot/model_test.go

```diff
@@ -0,0 +1,1243 @@
+		}
+
+		return nil
+	})
+	defer restore()
+
+	restore = boot.MockWriteModelToUbuntuBoot(func(model *asserts.Model) error {
+		writeModelToBootCalls++
+		switch writeModelToBootCalls {
+		case 1:
+			c.Assert(model, NotNil)
+			c.Check(model.Model(), Equals, "my-new-model-uc20")
+			// only 2 calls to reseal until now
+			c.Check(resealKeysCalls, Equals, 2)
+		case 2:
+			// handling of reseal with new model restores the old one on the disk
+			c.Check(model.Model(), Equals, "my-model-uc20")
+			m, err := boot.ReadModeenv("")
+			c.Assert(err, IsNil)
+			// and both models are present in the modeenv
+			currForSealing := boot.ModelUniqueID(m.ModelForSealing())
+			tryForSealing := boot.ModelUniqueID(m.TryModelForSealing())
+			// keys are first resealed for both models
+			c.Assert(currForSealing, Equals, "canonical/my-model-uc20,dangerous,"+s.keyID)
+			c.Assert(tryForSealing, Equals, "canonical/my-new-model-uc20,secured,"+s.keyID)
+
+		default:
+			c.Errorf("unexpected additional call to writeModelToBoot (call # %d)", writeModelToBootCalls)
+		}
+		return nil
+	})
+	defer restore()
+
+	err = boot.DeviceChange(s.oldUc20dev, s.newUc20dev)
+	c.Assert(err, ErrorMatches, `cannot reseal the encryption key: mock reseal failure`)
+	c.Assert(resealKeysCalls, Equals, 3)
+	c.Assert(writeModelToBootCalls, Equals, 2)
+	// still the old model file, all writes were intercepted
+	c.Check(filepath.Join(boot.InitramfsUbuntuBootDir, "device/model"), testutil.FileContains,
+		"model: my-model-uc20\n")
+
+	// finally the try model has been dropped from modeenv
+	m, err := boot.ReadModeenv("")
+	c.Assert(err, IsNil)
+	currForSealing := boot.ModelUniqueID(m.ModelForSealing())
+	tryForSealing := boot.ModelUniqueID(m.TryModelForSealing())
+	c.Assert(currForSealing, Equals, "canonical/my-model-uc20,dangerous,"+s.keyID)
+	// try model has been cleared
+	c.Assert(tryForSealing, Equals, "/,,")
+}
+
+func (s *modelSuite) TestDeviceChangeRebootRestoreModelKeyChangeMockedWriteModel(c *C) {
+	// system is encrypted
+	s.stampSealedKeys(c, s.rootdir)
+
+	oldKeyID := "Jv8_JiHiIzJVcO9M55pPdqSDWUvuhfDIBJUS-3VW7F_idjix7Ffn5qMxB21ZQuij"
+	newKeyID := "ZZZ_JiHiIzJVcO9M55pPdqSDWUvuhfDIBJUS-3VW7F_idjix7Ffn5qMxB21ZQuij"
+	// model can be mocked freely as we will not encode it
+	oldUC20dev := boottest.MockUC20Device("", boottest.MakeMockUC20Model(map[string]interface{}{
+		"model":             "my-model-uc20",
+		"brand-id":          "my-brand",
+		"grade":             "dangerous",
+		"sign-key-sha3-384": oldKeyID,
+	}))
+
+	newUC20dev := boottest.MockUC20Device("", boottest.MakeMockUC20Model(map[string]interface{}{
+		"model":             "my-model-uc20",
+		"brand-id":          "my-brand",
+		"grade":             "dangerous",
+		"sign-key-sha3-384": newKeyID,
+	}))
+
+	restore := boot.MockSeedReadSystemEssential(func(seedDir, label string, essentialTypes []snap.Type, tm timings.Measurer) (*asserts.Model, []*seed.Snap, error) {
+		s.readSystemEssentialCalls++
```

Comment:
(nitpick^2) I was initially a bit puzzled as this looks super similar to the original mocking but of course it uses a different model here. I suppose we could reuse the mocking by simply `s.oldUc20dev = boottest.MockU...` and `s.newUc20dev = ...`.